persistent/picklecache.py

```diff
@@ -229,36 +296,105 @@ def debug_info(self):
+            if node.object._p_state == UPTODATE:
+                # The C implementation will only evict things that are specifically
+                # in the up-to-date state
+                self._persistent_deactivate_ran = False
+
+                # sweeping an object out of the cache should also
+                # ghost it---that's what C does. This winds up
+                # calling `update_object_size_estimation`.
+                # Also in C, if this was the last reference to the object,
+                # it removes itself from the `data` dictionary.
+                # If we're under PyPy or Jython, we need to run a GC collection
+                # to make this happen...this is only noticeable though, when
+                # we eject objects. Also, note that we can only take any of these
+                # actions if our _p_deactivate ran, in case of buggy subclasses.
+                # see _persistent_deactivate_ran
+
+                node.object._p_deactivate()
+                if (self._persistent_deactivate_ran
+                    # Test-cases sneak in non-Persistent objects, sigh, so naturally
+                    # they don't cooperate (without this check a bunch of test_picklecache
+                    # breaks)
+                    or not isinstance(node.object, _SWEEPABLE_TYPES)):
+                    ejected += 1
+                    self.__remove_from_ring(node)
             node = node.next
+        if ejected:
+            # TODO: Only do this on PyPy/Jython?
+            # Even on CPython, though, it could trigger a lot of Persistent
+            # object deallocations and dictionary mutations
+            gc.collect()
+        return ejected
 
+    @_sweeping_ring
     def _invalidate(self, oid):
         value = self.data.get(oid)
         if value is not None and value._p_state != GHOST:
             value._p_invalidate()
             node = self.ring.next
-            while True:
-                if node is self.ring:
-                    break  # pragma: no cover belt-and-suspenders
+            while node is not self.ring:
                 if node.object is value:
-                    node.prev.next, node.next.prev = node.next, node.prev
+                    self.__remove_from_ring(node)
                     break
                 node = node.next
         elif oid in self.persistent_classes:
+            persistent_class = self.persistent_classes[oid]
             del self.persistent_classes[oid]
+            try:
+                # ZODB.persistentclass.PersistentMetaClass objects
+                # have this method and it must be called for transaction abort
+                # and other forms of invalidation to work
+                persistent_class._p_invalidate()
+            except AttributeError:
+                pass
+
+    def __remove_from_ring(self, node):
+        "Take the node, which previously contained a non-ghost, out of the ring"
```

Comment:
Docstrings don't look like docstrings when they're not delimited by the usual `"""..."""`.